### PR TITLE
Create the ion cannon

### PR DESCRIPTION
**Pull request by legokidlogan**

### CODE REVIEW

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+        Tracer = primary.TracerName == "" and 0 or primary.TracerFrequency,
+        Force = damage * 0.25,
+        Damage = damage,
+        Callback = function( _attacker, tr, dmginfo )
+            dmginfo:ScaleDamage( self:GetDamageFalloff( tr.StartPos:Distance( tr.HitPos ) ) )
+            dmginfo:SetDamageType( DMG_BULLET + DMG_DISSOLVE )
+        end
+    }
+
+    owner:FireBullets( bullet )
```
Comment (wrefgtzweve) [inline lua/weapons/cfc_ion_cannon/shared.lua:151]: This should be done in a different way, same for the trace in doExplosiveDamage.
https://wiki.facepunch.com/gmod/Entity:FireBullets
image
https://github.com/CFC-Servers/cfc_pvp_weapons/blob/main/lua/weapons/cfc_charge_gun_base.lua#L295-L301

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+SWEP.Slot = 4
+SWEP.Spawnable = true
+
+-- Appearance
+
+SWEP.UseHands = true -- If your viewmodel includes it's own hands (v_ model instead of a c_ model), set this to false
+
+SWEP.ViewModelTargetFOV = 65
+--SWEP.ViewModel = Model( "models/weapons/c_357.mdl" ) -- Weapon viewmodel, usually a c_ or v_ model
+--SWEP.WorldModel = Model( "models/weapons/w_357.mdl" ) -- Weapon worldmodel, almost always a w_ model
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:20]: These can be removed?

Also could you remove the default comments form this file as well

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+SWEP.ViewOffset = Vector( 0, 0, 0 ) -- Optional: Applies an offset to the viewmodel's position
+
+
+local doExplosiveDamage
+
+
+function SWEP:FireWeapon( chargeAmount )
+    local owner = self:GetOwner()
+    if not IsValid( owner ) then return end
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:155]: Is this possible?

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+        self._curChargeVMShakeStrength = vmFrac * primary.ChargeVMShakeStrength
+        self._curChargeVMShakeIntervalMin = Lerp( vmFrac, primary.ChargeVMShakeIntervalMinStart, primary.ChargeVMShakeIntervalMinEnd )
+        self._curChargeVMShakeIntervalMax = Lerp( vmFrac, primary.ChargeVMShakeIntervalMaxStart, primary.ChargeVMShakeIntervalMaxEnd )
+    end
+end
+
+function SWEP:OnOvercharged()
+    self:SetNextFire( CurTime() + self.Primary.Cooldown )
+    --self:SendWeaponAnim( ACT_VM_PRIMARYATTACK )
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:237]: Is this good to remove?

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+    eff:SetScale( 1 )
+    util.Effect( "Explosion", eff, true, true )
+
+    owner:EmitSound( "ambient/explosions/explode_4.wav", 80, math.Rand( 110, 115 ), 1 )
+
+    timer.Simple( 0, function()
+        if not IsValid( self ) then return end
+
+        self:Remove()
+    end )
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:270]: Does this actually need to be delayed? Remove should already queue the removal to the next frame

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+    timer.Create( timerName, 0.2, 0, function()
+        if not IsValid( self ) or not IsValid( self:GetOwner() ) then
+            timer.Remove( timerName )
+
+            return
+        end
+
+        self:SendWeaponAnim( ACT_VM_PULLBACK )
+    end )
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:300]: Interesting, why does it have to be like this? I think TankNut was telling us that timers in SWEPs are a code smell

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+
+    timer.Remove( "CFC_PvPWeapons_IonCannon_SpinAnim_" .. self:EntIndex() )
+end
+
+function SWEP:Reload()
+    self:SendWeaponAnim( ACT_VM_RELOAD )
+
+    local owner = self:GetOwner()
+
+    if IsValid( owner ) and owner:KeyPressed( IN_RELOAD ) then
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:321]: Could owner ever be invalid in this function?

File: lua/weapons/cfc_ion_cannon/shared.lua [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+        eff:SetOrigin( dmgPos )
+        eff:SetMagnitude( 2.5 * damageFrac )
+        eff:SetScale( 1.75 * damageFrac )
+        eff:SetRadius( radius * 0.35 )
+        eff:SetNormal( tr.HitNormal )
+        util.Effect( "AR2Explosion", eff, true, true )
+        util.Effect( "Sparks", eff, true, true )
+
+        local rf = RecipientFilter()
+        rf:AddAllPlayers()
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_ion_cannon/shared.lua:442]: I think you can AddPAS here instead

File: lua/weapons/cfc_trash_blaster/shared.lua [hunk 9]
```diff
@@ -187,6 +212,7 @@ function SWEP:FireWeapon( chargeAmount, notFirstCall )
     ) )
     prop:SetCollisionGroup( COLLISION_GROUP_INTERACTIVE_DEBRIS )
     prop:Spawn()
+    prop:SetOwner( owner )
```
Comment (brandonsturgeon) [inline lua/weapons/cfc_trash_blaster/shared.lua:215]: Does this also set CPPI? I forget

**Review by brandonsturgeon — APPROVED** [review]
Just a few questions, and some code cleanup 👍

File: lua/weapons/cfc_ion_cannon/shared.lua (hunk 3 above)
Comment (TankNut) [reply lua/weapons/cfc_ion_cannon/shared.lua:155]: Not under normal circumstances no, there are addons that can cause this but those are very specific.

File: lua/weapons/cfc_trash_blaster/shared.lua (hunk 9 above)
Comment (legokidlogan) [reply lua/weapons/cfc_trash_blaster/shared.lua:215]: It does not, I checked.

File: lua/weapons/cfc_ion_cannon/shared.lua (hunk 6 above)
Comment (TankNut) [reply lua/weapons/cfc_ion_cannon/shared.lua:300]: Yeah there is no reason to ever use a timer inside of a weapon, even if it's 'safe'. The ideal approach here is implementing the charge increment in think instead with it's own networkvar.

File: lua/weapons/cfc_trash_blaster/shared.lua (hunk 9 above)
Comment (brandonsturgeon) [reply lua/weapons/cfc_trash_blaster/shared.lua:215]: Should it?

Comment (legokidlogan) [reply lua/weapons/cfc_trash_blaster/shared.lua:215]: No, it not setting CPPI is intentional. :SetOwner() here just makes it not collide with the person shooting it so it doesn't interrupt their movement.

Comment (brandonsturgeon) [reply lua/weapons/cfc_trash_blaster/shared.lua:215]: Got it. Yeah we don't want to include them in some things, or give them the ability to use the trash later.

Addons that need it anyway should update their code to fallback check GetOwner I guess

**Pull request merged**